copy/copy.go

```diff
@@ -485,20 +497,58 @@ func (ic *imageCopier) copyLayers(ctx context.Context) error {
 		} else {
 			cld.destInfo, cld.diffID, cld.err = ic.copyLayer(ctx, srcLayer, pool)
 		}
-		data[index] = cld
+		data[index] = &cld
 	}
 
 	func() { // A scope for defer
 		progressPool, progressCleanup := ic.c.newProgressPool(ctx)
 		defer progressCleanup()
 
+		// digestToIndex records which index in `data` corresponds to
+		// which digest.  The intention behind is to avoid that we are
+		// copying the same blob multiple times which would yield a
+		// deadlock since `copyLayer()` is locking blobs which are
+		// unlocked at a very different part of the stack, after
+		// `Image()`.  Ideally `copyLayer()` would lock and unlock a
+		// given blob but that is not yet possible since
+		// `ImageDestination.PutBlob()` and `Commit()` are separated.
+		//
+		// digestToIndex works around the problem as it allows for
+		// checking if a given digest is being copied already.
+		digestToIndex := make(map[digest.Digest]int)
+
 		for i, srcLayer := range srcInfos {
-			copySemaphore.Acquire(ctx, 1)
+			_, ok := digestToIndex[srcLayer.Digest]
+			if ok {
+				// digest is already being copied
+				copyGroup.Done()
+				continue
+			}
+			// semMaxGoroutines protects us from malicious images
+			// with many layers trying to attack us.
+			ic.c.semMaxGoroutines.Acquire(ctx, 1)
+			digestToIndex[srcLayer.Digest] = i
 			go copyLayerHelper(i, srcLayer, progressPool)
 		}
 
-		// Wait for all layers to be copied
+		// Wait for all layers to be copied and propagate the copy
+		// data.
 		copyGroup.Wait()
+		for i, cld := range data {
+			if cld != nil {
+				continue
+			}
+			idx, ok := digestToIndex[srcInfos[i].Digest]
+			if !ok {
+				// Record an error if we have no entry in data
+				// AND no recorded index in digestToIndex.
+				data[i] = &copyLayerData{
+					err: errors.Errorf("copy data for blob %q should be cached but is not", srcInfos[i].Digest.String()),
+				}
+				continue
+			}
+			data[i] = data[idx]
```

Comment:
This is quite a bit of code only to keep the original semantics of `data`.  Why not make `data` a `map[digest.Digest]*copyLayerData` directly, and consume it directly below when building `destInfos` etc.?

That might require allocating a new `copyLayerData`, and setting `data[srcInfos[I].Digest]` in this goroutine, outside of `copyLayerHelper`, and passing the `copyLayerData` into the helper — but that does not look all that bad, it’s been non-obvious that the `data[index] = &cld` from concurrent goroutines is safe.

Reply:
Wonderful idea! I will follow your advise and refine that.

Reply:
Done.